Track ownership of scale subresource

nodo opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Track ownership of scale subresource for all scalable resources i.e. Deployment, ReplicaSet, StatefulSet, ReplicationController, and Custom Resources.

**Which issue(s) this PR fixes**:
Fixes #82046
Fixes #94585

**Special notes for your reviewer**:
I opened https://github.com/kubernetes/kubernetes/pull/95921 as an initial attempt to solve the problem. As this PR follows a different approach, I decided to open a new PR for clarity.

**Does this PR introduce a user-facing change?**:
```release-note
Track ownership of scale subresource for all scalable resources i.e. Deployment, ReplicaSet, StatefulSet, ReplicationController, and Custom Resources.
```

nodo committed:
Track ownership of deployments scale subresource

nodo commented:
/cc @apelisse @kwiesmueller

apelisse commented:
/assign
/triage accepted

kwiesmueller commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,224 @@
+		{
+			input: []metav1.ManagedFieldsEntry{
+				{
+					Manager:    "test",
```
you have the same manager twice here, I don't think that's valid for apply

nodo replied:
Makes sense, thanks for spotting!

kwiesmueller commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,224 @@
+		{
+			input: []metav1.ManagedFieldsEntry{
+				{
+					Manager:    "test",
```
same here, twice "test"

kwiesmueller commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,83 @@
+			return nil, err
+		}
+
+		updatedFieldSet := fieldset.Difference(toRemove)
```
Is this the kind of merge we need?
As far as I understand it always removes the provided path from the parent managedFields even if nothing was "stolen".

nodo replied:
Aah, good point, I will add this check and a unit test.

nodo replied:
I am not sure about this scenario though, when would this happen? And should we merge the managed fields of the subresource anyway?

kwiesmueller commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,224 @@
+			},
+		},
+		{
+			desc: "subresource path does not exist in parent entries",
```
could you also add a case where the path is not present on the subresource (so no ownership changes happened)?
At least that's what I think would happen in that case, right?

kwiesmueller commented:
/wg api-expression

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,83 @@
+
+	var filteredEntries []metav1.ManagedFieldsEntry
+	for _, entry := range entries {
+		if entry.FieldsV1 == nil {
```
It's kind of sad that you have to decide to decide that here. If we had more than one "Fields" version, we would have to remember to change that rather than using logic that probably/should exist somewhere else.

nodo replied:
The only thing that comes to mind to avoid this is to change `internal.FieldsToSet` to handle the case when its argument is `nil`. What do you think? Do you have something else in mind?

apelisse replied:
I suspect you're not supposed to use this function, but `fieldmanager.DecodeObjectManagedFields` instead, since it transforms the whole `ManagedFields` into the proper type (and you don't have to consider each individual field). Let me know if the return type doesn't give you what you need.

nodo replied:
It's definitely possible to use `DecodeObjectManagedFields` to check whether an entry has a certain field, but I am having hard time to change the entry to a specific version or change the field set. Do you have any thoughts on that?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,83 @@
+
+// FindEntries returns all the managed fields which has a specific path
+func FindEntries(entries []metav1.ManagedFieldsEntry, pathParts ...interface{}) ([]metav1.ManagedFieldsEntry, error) {
+	path, err := fieldpath.MakePath(pathParts...)
```
Can we build the path in the calling function so that the type is more obvious in the function signature (and doesn't require taking variable arguments)

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -329,6 +330,18 @@ func scaleFromDeployment(deployment *apps.Deployment) (*autoscaling.Scale, error
+	}
+	for i := range managedFieldsEntries {
+		managedFieldsEntries[i].APIVersion = "autoscaling/v1"
+		managedFieldsEntries[i].FieldsV1 = &metav1.FieldsV1{Raw: []byte(`{"f:spec":{"f:replicas":{}}}`)}
```
Building the field manually is quite unfortunate.

apelisse replied:
Also, I think we agreed for now that this would always be a constant final path (since that's an autoscaling field). Do we need to have the logic here where it would have to be repeated?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,83 @@
+// in case of scale) that needs to be removed by the managed field entries of
+// the parent.
+func MergeSubresourceEntries(parentEntries []metav1.ManagedFieldsEntry, subresourceEntries []metav1.ManagedFieldsEntry, APIVersion string, pathParts ...interface{}) ([]metav1.ManagedFieldsEntry, error) {
+	path, err := fieldpath.MakePath(pathParts...)
```
Are there cases where this path doesn't match the previous path (from FindEntries)? That seems repetitive (and error prone).

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,224 @@
+					Operation:  metav1.ManagedFieldsOperationApply,
+					APIVersion: "apps/v1",
+					FieldsType: "FieldsV1",
+					FieldsV1:   &metav1.FieldsV1{Raw: []byte(`{"f:spec":{"f:replicas":{},"f:selector":{},"f:template":{"f:metadata":{"f:labels":{"f:app":{}}},"f:spec":{"f:containers":{"k:{\"name\":\"nginx\"}":{".":{},"f:image":{},"f:name":{}}}}}}}`)},
```
I don't really care that the `FieldsV1` contains a real-life example since this is a unit-test. Here I suspect it makes things harder to parse for me as a reader. Have you considered building the fieldsV1 from a set instead?

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -404,5 +418,12 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
 	deployment.Spec.Replicas = scale.Spec.Replicas
 	deployment.ResourceVersion = scale.ResourceVersion
+
+	updatedEntries, err := fieldmanager.MergeSubresourceEntries(deployment.ManagedFields, scale.ManagedFields, "apps/v1", "spec", "replicas")
```
Same `deployment.MangedFields` here is passed again, and must ABSOLUTELY be the same as before. There is an implict state that should be preserved (including the `.spec.replicas` path) in a proper object.

nodo replied:
Totally, let me try and use your suggestion to organise the code a bit better.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,83 @@
+			return nil, err
+		}
+
+		updatedFieldSet := fieldset.Difference(toRemove)
```
Wait, I don't get how this algorithm works.

apelisse commented:
It's very encouraging that we're finally on the right track for this bug! I think there are a lot of improvements we can make to the code so that we don't fall into various traps (and most, if not all, my comments are related to that), but it's great that we're going the right direction!

liggitt commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -404,5 +418,12 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
 	deployment.Spec.Replicas = scale.Spec.Replicas
 	deployment.ResourceVersion = scale.ResourceVersion
+
+	updatedEntries, err := fieldmanager.MergeSubresourceEntries(deployment.ManagedFields, scale.ManagedFields, "apps/v1", "spec", "replicas")
```
rather than assuming `apps/v1` and `spec.replicas`, this should switch on the group/version of the request (defaulting to apps/v1 is ok for calls that don't have request info), something like:

https://github.com/kubernetes/kubernetes/blob/f4250198eac34e1adae7c627b9cb0cb4252bdb1e/pkg/registry/apps/deployment/strategy.go#L52-L56

If request info is found, but the group/version is not recognized, that should be a failure so we don't accidentally import apps/v1 fieldpaths into future versions automatically/incorrectly

liggitt replied:
ideally this could be encapsulated into a form where every scale subresource storage had a constant mapping of request group/version to fieldpath mapping, something like 

```go
mappings := map[schema.GroupVersion]mapping{
  schema.GroupVersion{Group:"apps",Version:"v1"}: mapping{
    From:[]string{"spec","replicas"}, // scale field path
    To: []string{"spec","replicas"}, // deployment field path
  },
}
```

or 

```go
mappings := map[schema.GroupVersion]mapping{
  schema.GroupVersion{Group:"acme.com",Version:"v1"}: mapping{
    From:[]string{"spec","replicas"}, // scale field path
    To: []string{"some","fieldPath","to","replicas"}, // v1 custom resource type field path
  },
  schema.GroupVersion{Group:"acme.com",Version:"v2"}: mapping{
    From:[]string{"spec","replicas"}, // scale field path
    To: []string{"some","other","fieldPath","to","replicas"}, // v2 custom resource type field path
  },
}
```

then select a mapping based on the request version, then call a helper, passing in the managed fields and mapping

nodo replied:
Thank you so much, makes sense to me.

I definitely like the idea of keeping a` mapping`, if I understand correctly the mapping is not constant right? because custom resource could define their own replicas path. In other words, there should be a way to "add" items to the mapping. Is that what you were thinking?

apelisse replied:
I don't know about the mapping to `.spec.replicas` in the scale object? That seems to be pretty fix, and doesn't depend on the type we're converting to/from. I think it's fine to have a map of type (Deployment, etc...) to replicas field.

nodo replied:
I addressed this in https://github.com/kubernetes/kubernetes/pull/98377/commits/79cdc00f8b65c46187a374d92cf8affccbebefb2, there is probably more refactoring we can do, but would be great if you could double-check that I am going in the right direction.

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -382,11 +370,25 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
 		return nil, errors.NewNotFound(apps.Resource("deployments/scale"), i.name)
 	}
 
+	managedFieldsHandler, err := fieldmanager.NewSubresourceHandler(
```
I don't see the path to `.spec.replicas` (within the deployment, not the scale object) here though?

nodo replied:
Good idea, I will make the change

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -17,50 +17,151 @@ limitations under the License.
+type SubresourceHandler struct {
+	parentEntries           []metav1.ManagedFieldsEntry
+	parentGroupVersion      schema.GroupVersion
+	subresourceGroupVersion schema.GroupVersion
```
You could probably drop this then.

apelisse replied:
You forgot to remove this :-)

nodo replied:
Mmmh I think we still need it, when we transform the take the managed fields of the main resource and transform them into scale we need to change there `APIVersion` into `subresourceGroupVersion` we can pass it in other ways though

nodo replied:
Added a field called `defaultGroupVersion` to clarify this.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -17,50 +17,151 @@ limitations under the License.
+		parentGroupVersion:      parentGroupVersion,
+		subresourceGroupVersion: subresourceGroupVersion,
+	}
+	m, ok := mappings[parentGroupVersion]
```
Take this as a parameter probably

nodo replied:
What are you referring to?

nodo replied:
Nvm I think I got it, the mappings belong to the storage and passed as parameter.

apelisse reviewed (commented):
I haven't looked in details, but that looks much better, thanks!

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,180 @@
+// resource and subresources
+type SubresourceScaler struct {
+	parentEntries           []metav1.ManagedFieldsEntry
+	parentGroupVersion      schema.GroupVersion
```
I'm pretty sure this is irrelevant, and depends on the version of each individual managedFields that you're looking at. Especially since the scaler here doesn't look at the object, but merely at the managedFields. You can probably remove this field.

nodo replied:
I have added it here because we need to convert back the updated managed fields into the correct API Version, I am not sure how to achieve that otherwise 🤔

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,180 @@
+// SubresourceScaler manages the conversion of managed fields between a main
+// resource and subresources
+type SubresourceScaler struct {
+	parentEntries           []metav1.ManagedFieldsEntry
```
The name is a little confusing I think. Maybe just `managedFields`?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,180 @@
+	parentGroupVersion      schema.GroupVersion
+	subresourceGroupVersion schema.GroupVersion
+
+	parentPath      fieldpath.Path
```
And so that depends on the specific version of each individual managed fields entry, so it has to be a "mapping".

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,180 @@
+}
+
+// ScalePathMappings maps a group/version to the replicas path in a scale subresource
+type ScalePathMappings map[schema.GroupVersion][]interface{}
```
I suspect the type could be: `map[schema.groupVersion]fieldpath.Path`

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,180 @@
+	// The assumption is that scale objects have autoscaling/v1 group version and
+	// corresponding .spec.replicas path
+	h.subresourceGroupVersion = schema.GroupVersion{Group: "autoscaling", Version: "v1"}
+	subresourcePath, err := fieldpath.MakePath("spec", "replicas")
```
I think it's fine to just replace this with `MakePathOrDie("spec", "replicas")`. If you don't want to die here, maybe pre-compute this in an init function, but there's really no reason for this to fail.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,180 @@
+	}
+	h.subresourcePath = subresourcePath
+
+	parentPath, err := fieldpath.MakePath(parentPathPieces...)
```
Compute the paths outside this function, and you probably need to receive a mapping. Maybe create the mapping directly to path so that they don't have to be recomputed (and checked for errors) everytime.

Since all this function does is compute these various things, I don't think it's super useful, people should be able to construct the object directly.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -47,16 +47,7 @@ func NewSubresourceScaler(parentEntries []metav1.ManagedFieldsEntry, parentGroup
-	if err != nil {
-		return nil, err
-	}
+	h.subresourcePath = fieldpath.MakePathOrDie("spec", "replicas")
```
That's much simpler. Do we still need this to be part of `h` since the value is always the same?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -35,10 +35,10 @@ type SubresourceScaler struct {
 
 // NewSubresourceScaler creates a new SubresourceScaler
-func NewSubresourceScaler(parentEntries []metav1.ManagedFieldsEntry, parentGroupVersion schema.GroupVersion, parentPathPieces []interface{}) (*SubresourceScaler, error) {
+func NewSubresourceScaler(parentEntries []metav1.ManagedFieldsEntry, parentGroupVersion schema.GroupVersion, parentPath fieldpath.Path) (*SubresourceScaler, error) {
```
I'm still pretty sure you need to take the full mapping rather than just `parentPath` here.

apelisse replied:
You still haven't addressed that comment.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,176 @@
+			}
+			entry.FieldsV1 = fields
+		}
+		entry.APIVersion = h.parentGroupVersion.String()
```
This is possibly changing the version that these fields are tracking, which is incorrect.

apelisse replied:
Each managedFields entry can map to a different version of the object.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+
+var (
+	scaleGroupVersion = schema.GroupVersion{Group: "autoscaling", Version: "v1"}
+	scalePath         = fieldpath.MakePathOrDie("spec", "replicas")
```
I'm assuming that's the path in the scale object? maybe `replicasPathInScale`?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+	t := map[string]*metav1.Time{}
+	for manager, versionedSet := range managed.Fields() {
+
+		if versionedSet.Set().Has(h.parentPath) {
```
`parentPath` depends on the version of the object (different versions can have the replicas path in different location), and you're not looking at the version of the deployment here.


This needs to look for the path in `replicasPathInDeployment` mapping for the version of that `versionedSet`.

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -60,10 +63,15 @@ func NewStorage(optsGetter generic.RESTOptionsGetter) (DeploymentStorage, error)
 	}
 
+	// maps a group version to the replicas path
+	scaleMappings := fieldmanager.ScalePathMappings{
```
this mapping is for within the deployment I'm assuming, not in the scale object? `replicasPathInDeployment`?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+	return managedFieldsEntries(internal.NewManaged(f, t))
+}
+
+func matchScaleManager(fields map[string]fieldpath.VersionedSet, otherManagerID string) (string, fieldpath.VersionedSet, bool) {
```
I can't see where this is used?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+
+	// 2. Decode the parent fields and process them to merge them with the
+	// subresource.
+	parentManagedFields, err := internal.DecodeObjectManagedFields(h.parentEntries)
```
It's a little sad that we have to decode them again here.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+	toRemove := fieldpath.NewSet(h.parentPath)
+	for manager, versionedSet := range parentManagedFields.Fields() {
+		// "steal" the scale path from the parent managed fields
+		newSet := versionedSet.Set().Difference(toRemove)
```
Since you don't check the version of the `versionedSet`, you don't know if that field that you remove exist for that specific version?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+		if versionedSet.Set().Has(scalePath) {
+			// TODO (nodo): handle multiple managed fields versions.
+			newAPIVersion := h.parentGroupVersion.String()
+			newVersionedSet := fieldpath.NewVersionedSet(
```
Isn't that replacing all the fieldsets that had the replicas field before? I don't think that's a good enough reason to remove it.

nodo replied:
I think with the new algorithm this is resolved.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+
+func managedFieldsEntries(entries internal.ManagedInterface) ([]metav1.ManagedFieldsEntry, error) {
+	obj := &unstructured.Unstructured{Object: map[string]interface{}{}}
+	if err := internal.EncodeObjectManagedFields(obj, entries); err != nil {
```
We should probably change that function signature if it requires us to do that weird dance here.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,192 @@
+
+// ToParent merges the `subresourceEntries` with the parent entries and
+// transform them accordingly
+func (h *SubresourceScaler) ToParent(scaleEntries []metav1.ManagedFieldsEntry) ([]metav1.ManagedFieldsEntry, error) {
```
Still not sure I agree with the algorithm here, here's what I can see:
1. There is a new value for replicas, whoever owned it before won't anymore, new manager is inserted
2. The value of replicas is the same as before, everyone should still be able to own their value
a. If the value was applied, we need a new manager for that apply, as in 1.
b. if the value was updated, I'm curious if we should add it anyway.

However, I don't know if that's the most convincing way to implement it.

Have you tried something like this:
- For each entry in parentsManagedFields that has replicas field,
  - Search corresponding entry in scaleManagedFields
    - If not found, remove the field from parent entry (it lost the field)
    - if found, remove that entry from scaleManagedFields (it kept the field)
- For each entry in scaleManagedFields left (assert that there should be one at most, newly added replicas owner)
  - Add new entry in parentsManagedFields of default version
- Return parentsManagedFields

nodo commented:
Rebased on top of https://github.com/kubernetes/kubernetes/pull/99210 to clean up the commit history.

desponda commented:
Hi 👋 Bug Triage here, just a reminder that code freeze is end of day today! PRs need to be merged in before then, or they will require an approved exception to go into the 1.21 milestone.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,151 @@
+// For the managed fields that have a replicas path it performs two changes:
+// 1. APIVersion is changed to the APIVersion of the scale subresource
+// 2. Replicas field of the main resource is transformed to `.spec.replicas`
+func (h *ScaleHandler) ToSubresource() ([]metav1.ManagedFieldsEntry, error) {
```
Wouldn't it be significantly easier to transform the data than the managed fields?

lavalamp commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -368,11 +376,25 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
+		return nil, err
+	}
+
 	// deployment -> old scale
```
Instead of constructing a scale object from the existing deployment, I expected us to construct a deployment object from the new scale object and the existing deployment object, and apply or update that?

nodo replied:
This is more or less what's happening. The problem is that in order to get the new scale object we need the "old scale" object from the deployment and run it through [UpdatedObject](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/apps/deployment/storage/storage.go#L378). UpdatedObject will update the managed fields of the scale object and conflict if needed.

lavalamp replied:
I think I am objecting to a large part of the existing codebase here. Antoine attempted to convince me that I have to hold my nose for this. I'll look again.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,151 @@
+// ResourcePathMappings maps a group/version to its replicas path
+type ResourcePathMappings map[string]fieldpath.Path
+
+// ScaleHandler manages the conversion of managed fields between a main
```
Since this code is in a very generic place, IMO it should be named for what it does, not what you intend to use it for. E.g. this is a "simple mapped-path handler" or something like that. (not sure if I'm asking for a change yet or not)

nodo replied:
True, the `struct` is potentially very generic. On the other hand we only manage scale throughout its logic. I think making more specific has value here because otherwise the number of edge cases explode.

lavalamp replied:
I also prefer keeping it specific--but then I'm not sure it should be located here?

apelisse replied:
We'll want to use it for CRDs and other types of resources. And we wanted to keep the "manager" manipulation within that package.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,151 @@
+// them so that they can be handled by scale.
+// For the managed fields that have a replicas path it performs two changes:
+// 1. APIVersion is changed to the APIVersion of the scale subresource
+// 2. Replicas field of the main resource is transformed to `.spec.replicas`
```
If this is a generic managed path mapping utility, best not to mention the specific field?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -0,0 +1,151 @@
+	t := map[string]*metav1.Time{}
+	for manager, versionedSet := range managed.Fields() {
+		path := h.mappings[string(versionedSet.APIVersion())]
+		if versionedSet.Set().Has(path) {
```
I think this assumes all the mappings refer to leaf fields, the comments should probably warn people.

nodo replied:
Makes sense, I added a comment where the `ResourcePathMappings` type is defined.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,621 @@
+					Operation:  metav1.ManagedFieldsOperationApply,
+					APIVersion: "apps/v1",
+					FieldsType: "FieldsV1",
+					FieldsV1:   &metav1.FieldsV1{Raw: []byte(`{"f:spec":{"f:replicas":{}}}`)},
```
Can you include the Subresource field on one of these?

nodo replied:
Done.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,621 @@
+					Operation:  metav1.ManagedFieldsOperationApply,
+					APIVersion: "apps/v1",
+					FieldsType: "FieldsV1",
+					FieldsV1:   &metav1.FieldsV1{Raw: []byte(`{"f:spec":{"f:selector":{}}}`)},
```
How do you know to remove the replicas field? It should only be removed if the value actually changes, no?

nodo replied:
Here "subresource" represent the managed fields of a Scale object _after_ it has been transformed from the old scale. 

Here is an example:
- Suppose there is a deployment with replicas set to 1, replicas is owned by manager `A`
- I apply an `autoscaling/v1` object with manager `B` and replicas set to 1 as well
- The field manager of scale will take it into account and update the managed fields of the scale object to have both `A` and `B` co-owning replicas
- We transform these managers back to the main resource. Note that this time `A` will appear as separate entry in the main resource because we added the "scale" field to the entry.

nodo replied:
If we are using `Update` instead of `Apply` it's the same, but we won't keep the two managers, only the one used to scale.

TLDR: it's up to the field manager of the scale object.

lavalamp replied:
In that case, shouldn't we have a test case with two managed fields from the subresource, where one matches up with an entry from the parent?

nodo replied:
Fair enough, let me add it.

nodo replied:
Added it [here](https://github.com/nodo/kubernetes/blob/scale-ownership-tracking-deployment/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go#L533)

apelisse replied:
Yeah, that test doesn't even look at the fields value, and it assumes that the fieldmanager update did the right thing. It's merely trying to convert the change made by the fieldmanager into deployments managedfields.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,621 @@
+					Operation:  metav1.ManagedFieldsOperationUpdate,
+					APIVersion: "autoscaling/v1",
+					FieldsType: "FieldsV1",
+					FieldsV1:   &metav1.FieldsV1{Raw: []byte(`{"f:spec":{"f:replicas":{}}}`)},
```
shouldn't the subresource field be set?

nodo replied:
Thanks for spotting, added.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,621 @@
+			},
+		},
+		{
+			desc: "Subresource steals all the fileds of the parent resource",
```
same question, actually on all of these -- how do you know the value actually changed and that co-ownership is no longer possible?

Jefftree commented:
/ok-to-test

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,696 @@
+					APIVersion:  "autoscaling/v1",
+					FieldsType:  "FieldsV1",
+					FieldsV1:    &metav1.FieldsV1{Raw: []byte(`{"f:spec":{"f:replicas":{}}}`)},
+					Subresource: "scale",
```
Why would this field be present?

apelisse replied:
Yeah, I don't think that case can happen. It looks like two scale managers have been added at the same time.
This suite of test is convenient to test the ToParent function specifically, but it works in such a vacuum that it's really easy to come-up with a scenario that can't really happen in practice, and this one is such a case.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource_test.go:
```diff
@@ -0,0 +1,696 @@
+					APIVersion:  "apps/v1",
+					FieldsType:  "FieldsV1",
+					FieldsV1:    &metav1.FieldsV1{Raw: []byte(`{"f:spec":{"f:replicas":{}}}`)},
+					Subresource: "scale",
```
I think this field wouldn't be present and therefore this manager is a duplicate, I think we shouldn't have duplicates in this list?

apelisse replied:
We still consider "test/apply/scale" to not be the same as "test/apply", so that applying on scale with an existing manager is not going to remove your previous applier.

lavalamp commented:
/approve

I'm not in love with this approach but the tests are good enough that I think it's best to take this.

lavalamp commented:
OK unit & integration tests passed.

/lgtm

lavalamp commented:
/milestone v1.21

lavalamp commented:
/lgtm

lavalamp commented:
/retest

can't tell if test is failing due to infra flake (can't find protoc??) or rebase problem.

lavalamp commented:
looks like openapi spec needed generation after that last rebase.

apelisse commented:
/retest

apelisse commented:
/lgtm

lavalamp commented:
/lgtm

nodo commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -3423,12 +3423,6 @@ func TestSubresourceField(t *testing.T) {
-		managedFields[2].Subresource != "status" ||
-		string(managedFields[2].FieldsV1.Raw) != `{"f:status":{"f:unavailableReplicas":{}}}` {
-		t.Fatalf(`Unexpected entry, got: %v`, managedFields[1])
-	}
```
What was the reason to remove this?

nodo replied:
Oh I see I have forgotten about adding a call to scale as well, I think it would be nice to keep both

nodo commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go:
```diff
@@ -113,17 +113,24 @@ func (h *ScaleHandler) ToParent(scaleEntries []metav1.ManagedFieldsEntry) ([]met
+			// Field wasn't stolen, let's keep the entry as it is.
+			f[manager] = versionedSet
 			t[manager] = decodedParentEntries.Times()[manager]
+			delete(scaleFields, manager)
```
I am not sure I fully understand this change. The keys of  scaleFields will always have the `-scale` suffix so the parent manager will match only i it has the same manager name and it has Subresource field set to scale. In this case, it's OK to steal I think. The result is the same but the code is more complicated.

lavalamp replied:
No, because there could be a remaining entry from when we extracted entries from the parent resource in the first place.

nodo replied:
right, agree, it's an edge case that I missed.

nodo commented:
/retest

nodo commented:
/retest

apelisse commented:
/lgtm

apelisse commented:
/test pull-kubernetes-integration

apelisse commented:
/hold

Let's wait for Joe's PR to be merged first.

liggitt commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -376,11 +383,22 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
 
+	managedFieldsHandler := fieldmanager.NewScaleHandler(
+		deployment.ManagedFields,
+		schema.GroupVersion{Group: "apps", Version: "v1"},
```
apps/v1 happens to be the only requestable version currently, this should really get the group/version from the request info in the context rather than assuming

that should be done as a follow-up, not in this PR

nodo replied:
Yes, I have forgotten to re-add it. I will create a separate PR for that.

nodo committed:
Check request info when updating managed fields during scale

- Test all versions to make sure each resource version is in the
  mappings
- Fail when request info contains an unrecognized version. We have tests
  that guarantee that all known versions are in the mappings. If we
  get a version in request info that is not there we should fail fast to
  prevent inconsistent behaviour (e.g. for some reason the mappings is
  not up to date).

Ensure all known versions are in mappings

apelisse committed:
Use ScaleHandler for all scalable resources

apelisse commented:
/hold cancel

apelisse commented:
/retest

nodo commented:
Rebased and re-added generated code.

nodo commented:
/retest

nodo commented:
/retest

annajung commented:
Hi all, Clearing the milestone as @apelisse informed me that this is not going to make it into 1.21.
/milestone clear

nodo committed:
Do not add managed fields if a scale entry doesn't own replicas

This happens when a request changes the .status.replicas but not
.spec.replicas

jpbetz commented:
`TestApplyScaleSubresource` failure appears to be due to this PR: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98377/pull-kubernetes-integration/1378757700921331712

nodo commented:
/retest

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -376,11 +384,30 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
 		return nil, errors.NewNotFound(apps.Resource("deployments/scale"), i.name)
 	}
 
+	defaultGroupVersion := schema.GroupVersion{Group: "apps", Version: "v1"}
```
Not sure default is the right word here. It's the default value until it's not.

apelisse replied:
Oh yeah, actually it's the default. Forget about this comment :-)

apelisse reviewed (commented):
Mostly looks good to me. Can you squash some of these commits together maybe?

liggitt commented on pkg/registry/apps/deployment/storage/storage_test.go:
```diff
@@ -284,6 +284,83 @@ func TestScaleUpdate(t *testing.T) {
+
+	var deployment apps.Deployment
+	ctx := genericapirequest.WithNamespace(genericapirequest.NewContext(), namespace)
+	ctx = genericapirequest.WithRequestInfo(ctx, &genericapirequest.RequestInfo{
```
yeah... this multi-version-with-different-replicas-paths scenario seems much more relevant to add for a multi-version custom resource, not deployments

apelisse replied:
Here I think it's specifically trying to see what happens when we end-up with a requestinfo that has a version we don't know about. Not sure that's useful. There is already a multi-version CRD test in that PR somewhere else.

liggitt replied:
well... now that I think about it... if a scaleable resource ever added a newer apiVersion, knowing what an n-1 API server would do is important to be able to upgrade confidently.

apelisse replied:
But how would we even end-up here? By doing `apps/v2/deployment/.../scale` with apps/v2 deployment that doesn't exist? I don't think the end point would even be setup

liggitt replied:
* 1.30 adds apps/v2 deployment
* HA cluster has some 1.29 API servers, some 1.30 API servers
* User does apps/v2/deployment/.../scale update via 1.30 API server
* User does apps/v1/deployment/.../scale update via 1.29 API server, which then has to interpret the apps/v2 managedFields entry

apelisse replied:
Yeah, the 1.29 (is that a prediction? I'll take notes ;-) server will probably drop all the v2 entries anyways, but I don't think that's what's going on here.

nodo replied:
I still think these tests are useful though, would you like me to change them or remove them?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/scalehandler.go:
```diff
@@ -0,0 +1,163 @@
+	f := fieldpath.ManagedFields{}
+	t := map[string]*metav1.Time{}
+	for manager, versionedSet := range managed.Fields() {
+		path := h.mappings[string(versionedSet.APIVersion())]
```
Since @liggitt mentioned it, clearly the mappings here might miss a version as we update the apiserver. What do you think should the behavior be in that case? For normal resource, we typically entirely remove the entry that doesn't have a corresponding version.

nodo replied:
Here we were dropping the field anyway because "path" was empty string. I have added a check to make it explicit: https://github.com/kubernetes/kubernetes/pull/98377/commits/55c3ec0b06c2eed475e266353417d607ad0d413d

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/scalehandler.go:
```diff
@@ -0,0 +1,163 @@
+
+	for manager, versionedSet := range parentFields {
+		// Get the main resource "replicas" path
+		path := h.mappings[string(versionedSet.APIVersion())]
```
I think here you would have to drop the entry if the mapping doesn't exist for that version.

nodo replied:
Added it in https://github.com/kubernetes/kubernetes/pull/98377/commits/55c3ec0b06c2eed475e266353417d607ad0d413d with unit tests, WDYT?

nodo committed:
Drop managed fields entries with unknown fields

This is aligned to the behaviour of server-side apply on main resources.

nodo commented:
I have rebased and cleaned up the commit history.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/scalehandler.go:
```diff
@@ -68,7 +68,12 @@ func (h *ScaleHandler) ToSubresource() ([]metav1.ManagedFieldsEntry, error) {
-		path := h.mappings[string(versionedSet.APIVersion())]
+		path, ok := h.mappings[string(versionedSet.APIVersion())]
+		// Drop the field if the APIVersion of the managed field is unknown
+		if !ok {
```
My concern with this is that we will remove entries for mappings that don't exist, not really for versions that don't exist. What if forgot to create a mapping? How would we know other than because this triggers a bug of managedFields being removed?

nodo replied:
This is totally a valid point, I am not sure how to check that though 🤔

apelisse replied:
OK I've created a test for `ReplicaSet`:
```golang
func TestReplicasForAllVersions(t *testing.T) {
	knownVersions := scheme.Scheme.VersionsForGroupKind(schema.GroupKind{Group: "apps", Kind: "ReplicaSet"})
	for _, version := range knownVersions {
		if _, ok := replicasPathInReplicaSet[version.String()]; !ok {
			t.Fatalf("missing version %v in `replicasPathInReplicasSet`", version)
		}
	}

	if len(knownVersions) != len(replicasPathInReplicaSet) {
		t.Fatalf("`replicasPathInReplicaSet` has extra items: %v vs %v", replicasPathInReplicaSet, knownVersions)
	}
}
```

It's failing because we don't have apps/v1beta Replicaset, looks like this was useful :-)

apelisse replied:
I used `k8s.io/client-go/kubernetes/scheme` for the scheme by the way.

nodo replied:
Thanks a lot @apelisse ! 🎉  I have add the tests in https://github.com/kubernetes/kubernetes/pull/98377/commits/cc220b6ea4cdcff66d415be2a925bb88367b76c1

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -165,6 +165,11 @@ func (c *crConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVe
 		// TODO: should this be a typed error?
 		return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", fromGVK.String(), target)
 	}
+	scaleGVK := schema.GroupVersionKind{Group: "autoscaling", Version: "v1", Kind: "Scale"}
```
are you sure? we explicitly handle this in https://github.com/kubernetes/kubernetes/blob/7ca1037d8b2112bfe190e2493611296518380e73/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go#L123-L131

nodo replied:
yeah, I think this bit was missing to handle scale properly.

This is what happens:
1. `toVersioned` calls `ConvertToVersion` on the converter object [here](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/structuredmerge.go#L168)
2. `validVersions` won't contain the `autoscaling/v1`, and [this check](https://github.com/kubernetes/kubernetes/blob/7ca1037d8b2112bfe190e2493611296518380e73/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go#L168) fails

So we won't call `Convert` and fail early.

liggitt replied:
yeah, teaching this about scale conversion is fine. do it like this so we only make the exception when there's actually a scale subresource, and so we only make the exception for actual typed Scale objects (which means we're coming in via the scale subresource request and not doing something wacky like trying to submit autoscaling/v1 Scale to the main resource endpoint)

```go
	// Special-case typed scale conversion if this custom resource supports a scale endpoint
	if c.convertScale {
		if _, isInScale := in.(*autoscalingv1.Scale); isInScale {
			return typedscheme.Scheme.ConvertToVersion(in, target)
		}
	}
```

nodo replied:
Makes sense, thanks 👍

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -383,9 +384,17 @@ func (i *scaleUpdatedObjectInfo) UpdatedObject(ctx context.Context, oldObj runti
 	}
 
+	defaultGroupVersion := schema.GroupVersion{Group: "apps", Version: "v1"}
+	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found {
```
That's great, I'm just curious how we would end-up with a `requestInfo` that doesn't know about a version we're serving here. Also I think we should rename `defaultGroupVersion` to merely `groupVersion` since default is a little misleading.

nodo replied:
I have renamed the variable and cleaned up the commit history.

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -700,6 +700,27 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
+			utilruntime.HandleError(err)
+			return nil, fmt.Errorf("the server could not properly serve the CR subresources")
+		}
+		if subresources == nil || subresources.Scale == nil {
```
But is the scale handler going to gently ignore these, or is it going to remove their entire entry?

apelisse replied:
We absolutely have to fix that.

nodo replied:
That's a good point. 

A possible solution would be for the scale handler to always ignore entries with version it doesn't know about, instead of dropping. This would apply for both CRs and other resources.  That would be inconsistent with `Apply`, I think.

Another way would be to somehow limit this behaviour to CRs, by passing some parameter down to the scale handler, so that we have a different logic for custom resources. I am not sure I like that though.

apelisse replied:
I think there is a difference between versions that don't exist and versions that don't have a scale subresource.

nodo replied:
Right, what about https://github.com/kubernetes/kubernetes/pull/98377/commits/1f1d9387ef0d3051d560b4c2bcbdc6bb7d45db60. 

Basically, I am adding an empty path to all CR versions that don't have `scale`. The assumption is that manage fields entries won't ever have an empty path, so those entries just gets ignored [here](https://github.com/nodo/kubernetes/blob/scale-ownership-tracking-deployment/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/scalehandler.go#L121-L124)

apelisse commented:
You can remove a few commits when you rebase! Thanks :-)

nodo committed:
Add nil path to mapping when a CR has no "scale" subresource

This is to prevent the ScaleHandler to drop the entry. In this way
entries just get ignored.

nodo commented:
Rebased and cleanup the commit history.

nodo commented:
/retest

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/scalehandler.go:
```diff
@@ -0,0 +1,174 @@
+		if !ok {
+			continue
+		}
+
```
Maybe special treat the case where `path` is `nil` here, so that you conserve the entry without ever stealing anything from it.

nodo replied:
Yes much better.

At first I excluded to use `nil` because I didn't want to enforce `Path` to be something "nullable", e.g. if we want to change it to a struct in the future. I think it's OK to use nil for now though :-) 

I force-pushed here: https://github.com/kubernetes/kubernetes/commit/6c86e89390e262cc54be647bc09328a4d8d654fb

apelisse commented:
Looks good to me, I'm curious about what we can do for the conversion, a small nit on nil path and I think that's good to go.

/assign @liggitt

apelisse commented:
I think the only question left is the conversion! Thank you :-)

lavalamp commented on pkg/registry/apps/deployment/storage/storage_test.go:
```diff
@@ -39,6 +40,7 @@ import (
 	"k8s.io/apiserver/pkg/registry/rest"
 	storeerr "k8s.io/apiserver/pkg/storage/errors"
 	etcd3testing "k8s.io/apiserver/pkg/storage/etcd3/testing"
+	"k8s.io/client-go/kubernetes/scheme"
```
yeah this isn't right, although it's only a test so maybe it doesn't matter. But there should be something similar in an apiserver subpackage somewhere.

nodo replied:
As you suggested below, I separated the tests into `storage_test` in https://github.com/kubernetes/kubernetes/pull/98377/commits/f8a6ee7cb7ac3d958714028ecaece02463e298fb. It has the side effects that we have to export the mappings variables. WDYT?

apelisse replied:
Not a big fan of exporting the mapping :-/ Do dependencies in `_test.go` files really affect the dependency tree for people who import the package?

nodo replied:
Moved it to the integration tests.

lavalamp commented:
I don't think registry should be importing the client, if you have to can you at least change the package name (add "_test") so that the import graph isn't weird?

lgtm other than that.

nodo commented:
/retest

apelisse commented on pkg/registry/apps/deployment/storage/storage.go:
```diff
@@ -58,7 +58,7 @@ type DeploymentStorage struct {
 
 // maps a group version to the replicas path in a deployment object
-var replicasPathInDeployment = fieldmanager.ResourcePathMappings{
+var ReplicasPathInDeployment = fieldmanager.ResourcePathMappings{
```
Making it public is fine, but it should be returned by a function so that people can't modify it. And let's move the test to integration, thanks!

nodo replied:
Makes sense, will amend the commit

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/structuredmerge.go:
```diff
@@ -165,17 +165,9 @@ func (f *structuredMergeManager) Apply(liveObj, patchObj runtime.Object, managed
 
 func (f *structuredMergeManager) toUnversioned(obj runtime.Object) (runtime.Object, error) {
-	// QUESTION(nodo): That seems reasonable I think, but not clear if this is the right fix yet
-	if obj.GetObjectKind().GroupVersionKind().GroupVersion() == f.groupVersion {
```
would be good to squash the fix into this so we don't have an intermediate commit with this... this is really confusing to see in the commit history

nodo replied:
Sounds good.

nodo commented:
The last commit is still WIP, I will squash it with a previous one.

nodo commented:
Squashed the commit and rebased.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -165,6 +165,13 @@ func (c *crConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVe
 		// TODO: should this be a typed error?
 		return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", fromGVK.String(), target)
 	}
+	// Special-case typed scale conversion if this custom resource supports a scale endpoint
```
lgtm

liggitt reviewed (commented):
conversion change lgtm

apelisse commented:
OK, Thanks @nodo!

/lgtm

Who wants to approve, @lavalamp @liggitt ?

liggitt commented:
> Who wants to approve, @lavalamp @liggitt ?

Daniel's approval still holds

apelisse commented:
Ah, I missed it, thank you :-)

apelisse commented:
/retest

nodo commented:
/retest

nodo commented:
/retest
